### ISSUE TITLE
BPMN support?

### ISSUE THREAD

**Issue by windischb** [elsa-workflows/elsa-core#39]
I'm not sure if  Elsa supports all necessary functions, but have you ever thought about supporting BPMN Models?
BPMN Models are simple XML Files...

I think this could be a huge benefit because of interportibility!
Using an existing BPMN Model( which for example currently runs in Camunda), change a few Parameters, and run this Model in Elsa would be amazing!

Do you think thats possible?

Currently i'm not sure where to start for trying to implement an "Elsa BPMN Parser"...
I will try to find out, how, or if this can be done with Elsa.

If you can give me any directions this would be great!

**Comment by sfmskywalker** [comment]
Yes! It’s been a long-standing wish to have BPMN interoperability. I don't think I have a nice extensibility endpoint atm to translate between formats, but I do intend to fix that. This would totally allow you to implement e.g. a BPMN import/export function. Elsa doesn’t support everything yet that is in the BPMN spec, but we could work in that.

You're more than welcome to contribute of course. However I’m in the middle of a big cleanup (see “simplify” branch), so I think it’s best advised for that branch to get merged in. Hopefully after the weekend.

**Comment by windischb** [comment]
some time has passed ...
I think I finally have enough time to start implementing or at least try to implement the first parts for the BPMN functionality ...
How do you think it should work?
Convert the BPMN structure to JSON, which can then be imported?
Given lost information from import, this would likely be a one-way import.
Or would it be conceivable to expand the classes in Elsa with the missing properties in order to be able to export the whole thing back to BPMN?

**Comment by hidcontrol** [comment]
maybe [it](https://forum.camunda.org/t/camunda-xml-json-xml/18888/12) will be useful

**Comment by zhouguoqing** [comment]
Support BPMN will be powerful. @sfmskywalker

**Comment by tunechr** [comment]
Just MO here, BPMN is very powerful, but it becomes really complex and is an art in it's self. 
 
I can see that under the hood ELSA could support BPMN, or import it...  Just check out the spec:
https://www.omg.org/spec/BPMN/2.0/PDF

What I like about ELSA is the focus on an easy/slick UI and I'm guessing the market here is low code automation where it's easy to use by 'normal' people....something that BPMN might not be. 

The BPMN 2.0 spec means you might need to have the BPMN diagram notation as well, so using something like https://bpmn.io/ which at least you have a ready-made designer!  But it is abstracted from anything business domain related. 

If you look at Microsoft Flow, you can import BPMN and create flows (with some restrictions), but the whole flow thing gets messy. 

I'm just throwing this out there, I just feel BPMN might not be ELSA's true north.

If you want BPMN you might just go to something like **Camunda** which is open source and has a stack of other products (although it is java there are some .net client connectors).  

Feel free to ignore me!

**Comment by craigfowler** [comment]
There are many dupes of this which might have worthwhile discussion.  There is also a discussion thread at #639 .  I've assigned it low priority as it doesn't seem to be on any milestones right now.  But "low priority" certainly _does not have to mean_ "never".

**Comment by Varorbc** [comment]
@sfmskywalker Is there any update?

**Comment by sfmskywalker** [comment]
For BPMN support, we need Elsa 3 which is designed to support different kinds of workflows. Initially it comes with a simple Flowchart and Sequence as diagram types. Once it’s on par with Elsa 2, we might start to think about a BPMN diagram type. But, as always, it depends on how it’s priority relates to those of other features and what our bandwidth looks like.

**Comment by fromz** [comment]
@craigfowler @sfmskywalker am I right in understanding that the project aims to release 3.0 first, and then consider how we might approach introducing BPMN to the project? or are we looking at BPMN support /for/ the 3.0 release? do we have any gut feel as to how far off 3.0 might be?

**Comment by sfmskywalker** [comment]
That's exactly right; the initial 3.0 release will be **without** BPMN support. The architecture, however, allows for adding BPMN support later on.

The way things are looking now, the 3.0 release candidate will happen around the end of this year/beginning of next year. I already started work on documenting 3.0, which helps me to identify important missing pieces before the release.

**Comment by sfmskywalker** [comment]
BPMN web modeler: https://bpmn.io/toolkit/bpmn-js/

**Comment by ThatDeveloper** [comment]
@sfmskywalker Hey there! I would love to see BPMN support, it is the only thing that currently prevents us from switching to ELSA. We love ELSA, but without BPMN it would mean additional work to translate processes, and it outweighs the benefits for us.

**Comment by barnybadzoo** [comment]
Is there an timeline when BPMN support is available ? We also would love to use Elsa in our project but the missing bpmn support prevents this. We still got a bit time, so i wanted to ask you for current plans of the implementation.

**Comment by sfmskywalker** [comment]
Our current plan is to start working on this in Q4, contingent on our reliability & performance deliverables.

**Comment by barnybadzoo** [comment]
thank you for that info.. hopefully you will also provide a migration of existing workflows. that would be great😁

**Comment by barnybadzoo** [comment]
any updates on this ? would be great to see BPMN soon.

**Comment by ThatDeveloper** [comment]
Agreed. BPMN and DMN is what would make this the killer library for us. We would appreciate it.

**Comment by sfmskywalker** [comment]
I'm afraid that the BPMN feature has been deprioritized and is not currently part of the 2025 roadmap. While we understand its importance, we’ve had to prioritize other features based on available resources.

**Comment by mingyangzhu** [comment]
BPMN and DMN is  importance，Once implemented, the scope of use is greatly enhanced. Very helpful for ecological construction.

**Comment by barnybadzoo** [comment]
This post is very old. I'am waiting for a long time for this feature, and now you had deprioritized this, and we even can't expect this feature even in 2025 ??  Why did you drop this ? 

We wanted to be able to use BPMN directly for managing our workflows. Now we  gonna swtch to an alternative..  Check here: https://github.com/meirwah/awesome-workflow-engines

Without a clear Roadmap, we will not use elsa workflow for our purpose

**Comment by svenhue** [comment]
Vote this comment if you would use elsa with bpmn support.
I am currently working with Camunda but seriously interested implementing bpmn 2.0 in elsa workflows so please give me feedback about if this would be valuable for you.

As @sfmskywalker already mentioned, the architecture allows an bpmn extension. 
Can you tell about the current state about implementing bpmn? Looks like this feature is in backlog and the dev is not planned/ started?

**Comment by sfmskywalker** [comment]
@svenhue Upvoted 👍🏻 

No work has been done yet on BPMN support, so it’d be awesome if you’re up for tackling it.

Before jumping in, it might be worth sketching out your approach in technical terms. For instance, are you planning to lean on the bpmn-js library from Camunda (which seems like a solid choice)? If so, how would that fit with the existing Elsa activities? Speaking of activities, I’m picturing a new type—maybe something like `BpmnDiagram`—as a third container activity alongside `Flowchart` and `Sequence`. This one would handle interpreting and running BPMN models. Since BPMN models are made up of activities themselves, would it make sense to map those directly to Elsa activity classes? Like, could a BPMN "XOR" activity translate to something like an Elsa `BpmnXor` class? Or does that not really work, and you’d rather build a totally separate set of classes for BPMN activities? Just some food for thought - curious to hear your take!

**Comment by svenhue** [comment]
@sfmskywalker 

I didnt thought about the frontend/process modelling part of the implementation until now, i would start with type definitions for bpmn 2.0 and then debug the existing workflow execution flow to get an understanding of how elsa works internally and how the bpmn extension has to be implemented. My target would be to understand the differences and commonalities of the technical concepts. So to be clear: wouldnt be the first step to implement the hole bpmn2.0 definition (metadata, classes) without dependencys on elsa and then to elaborate how this concepts can be merged in way that we can use the existing workflow executions logic and all that existing awesome stuff  (Elsa.Workflows.Core etc.)? So then i have a new class library called Elsa.Bpmn and then i try to implement interfaces/ inherit classes of elsa so that we dont need to change anything of e.g Elsa.Workflows.Core/services/WorkflowRunner. the approach with BpmnDiagram that inherits Container and class BPMNTask that inherits Activity seems also reasonable.  BPMNXor - what i see here is that elsa has slightly different concept of activitys - in bpmn activitiys, event and gateways are flow objects - in elsa an event is a activity (?)  ... I will go for trail and error first, then commit my changes in my fork and then we will elaborate how it goes:)

**Comment by sfmskywalker** [comment]
@svenhue 

Yep, those are all great questions - love how you’re digging into the nitty-gritty! Your trial-and-error plan sounds perfect to me. Starting with type definitions and debugging Elsa’s internals to figure out the BPMN fit makes a lot of sense. I’m excited to see how it shapes up in your fork - keep me posted on how it goes! :)

**Comment by omid-tansaz** [comment]
> Just MO here, BPMN is very powerful, but it becomes really complex and is an art in it's self.
> 
> I can see that under the hood ELSA could support BPMN, or import it... Just check out the spec: https://www.omg.org/spec/BPMN/2.0/PDF
> 
> What I like about ELSA is the focus on an easy/slick UI and I'm guessing the market here is low code automation where it's easy to use by 'normal' people....something that BPMN might not be.
> 
> The BPMN 2.0 spec means you might need to have the BPMN diagram notation as well, so using something like https://bpmn.io/ which at least you have a ready-made designer! But it is abstracted from anything business domain related.
> 
> If you look at Microsoft Flow, you can import BPMN and create flows (with some restrictions), but the whole flow thing gets messy.
> 
> I'm just throwing this out there, I just feel BPMN might not be ELSA's true north.
> 
> If you want BPMN you might just go to something like **Camunda** which is open source and has a stack of other products (although it is java there are some .net client connectors).
> 
> Feel free to ignore me!


---

Very fair points and a great discussion in this thread.  

We have extensive real-world experience using a BPMN engine (Camunda) in a high-throughput, complex insurance platform. We believe we can contribute valuable insights to this discussion:  

1. **BPMN 2.0 as a Standard**  
   BPMN 2.0 is a standard, and fully implementing it can be a challenging and lengthy process, as it requires adherence to OMG specifications. We have firsthand experience with this through our close partnership with Camunda. It took Camunda multiple iterations to integrate (almost) the full BPMN 2.0 specification into Zeebe (Camunda 8), highlighting the complexity of such an undertaking. These challenges should not be underestimated.  

   That said, this does not mean we cannot take a similar iterative approach—gradually incorporating more BPMN specifications over time. However, this requires alignment and collaboration to prioritize which specifications should be implemented first. We can learn a lot from the path Camunda has taken, given their deep involvement in the OMG and BPMN communities.  

2. **Where BPMN Makes Sense**  
   It’s important to assess the real benefits of BPMN and where it makes sense to apply it. We have been using BPMN 2.0 with Camunda in combination with Elsa for over seven years. There are distinct differences between BPMN and Elsa workflows in terms of use cases.  

   The most significant difference is that BPMN (Camunda) is best suited for **business process orchestration**—that is, handling high-level business workflows that require visibility and abstraction. However, for **low-level technical automation**, BPMN can become restrictive, particularly if one aims to follow best practices in BPMN design.  

   From our experience, the most effective approach has been to combine a **business process orchestration engine (Camunda, using BPMN)** with an **execution-level workflow engine (Elsa) for detailed technical automation**.  

   Of course, this adds architectural complexity and requires significant maintenance to run both engines.  

3. **Potential for BPMN Support in Elsa**  
   If Elsa were to support BPMN execution while adhering to BPMN standards and diagrams, we could unify **both business process orchestration and technical workflows within a single technology (Elsa).**  

   - BPMN workflows would empower **business users** to design business process orchestrations visually.  
   - Elsa workflows would provide **automation engineers** full flexibility to design complex execution-level workflows without the constraints of BPMN.  

If this idea resonates with others, we would be happy to start a more in-depth discussion. We can also document our implementation at nexxbiz and share insights on what has worked well and what challenges we have encountered. Perhaps this could help inform the design discussions moving forward.

**Comment by basouzars** [comment]
any news?

**Comment by sfmskywalker** [comment]
BPMN support is coming to Elsa 4.0 (under development).

Early alpha screenshot:

<img width="2560" height="1352" alt="Image" src="https://github.com/user-attachments/assets/a15ce387-1771-47d4-8b37-d55c0cbe8bb1" />

**Comment by sfmskywalker** [comment]
Forward reference for anyone landing here from a search: I have written up a full design for BPMN 2.0 support in Elsa 3 as a response to this request, at #7909.

Short version. The BPMN semantics now exist as a standalone MIT-licensed .NET library with no external NuGet dependencies (`valence-works/bpmn`), so Elsa does not have to implement token semantics, gateways or boundary events itself. The design covers how that library plugs into Elsa 3, whether BPMN elements should be Elsa activities (recommendation: no, one composite activity plus an interpreter, with an inspection projection on top), designer integration including the bpmn-js license question, a v1 scope, a work breakdown, risks and open questions.

I also ran a throwaway spike against `release/3.8.0` before writing it. Three of four test processes pass and express all three of the library's host commands. The fourth failed, and that failure is the useful part: Elsa 3 has no way for a container activity to be told that a child faulted, so BPMN error boundary events need a new seam in `Elsa.Workflows.Core`. Details in section 2.1 of #7909.

Nothing is implemented and no PR is open. #7909 needs a decision before any of that happens, so comments belong there rather than here.

**Comment by sfmskywalker** [comment]
BPMN 2.0 support has shipped for Elsa 3, following Camunda's lead on what is executable. The design is #7909 and its closing comment covers the full scope.

- **Runtime:** a `BpmnProcess` activity. It imports BPMN 2.0 XML (including Camunda-authored files) through `POST bpmn/analyze` / `bpmn/import`, binds tasks to Elsa activities, and runs boundary events, gateways, error propagation, compensation, transactions, event subprocesses and multi-instance.
- **Studio:** a read-only BPMN designer with import/export, a Performed-by panel to bind tasks to activities, an instance overlay per element, and drill-down from a Flowchart into a nested BPMN process.
- **Docs:** `doc/wiki/bpmn-workflows.md`.

Deliberately later: visual BPMN editing (elsa-workflows/elsa-studio#1004) and a React Flow adapter (elsa-workflows/elsa-studio#1003). Known library positions, such as expression-based flow conditions not being evaluated, are listed on #7909.
